torch/distributed/_tensor/api.py

```diff
@@ -293,6 +299,8 @@ def __coerce_same_metadata_as_tangent__(self, metadata_tensor):
     # pyre-fixme[3]: Return type must be annotated.
     # pyre-fixme[2]: Parameter must be annotated.
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        if func in metadata_handlers:
+            return metadata_handlers[func](*args, **kwargs)
```

Comment:
What are the chances this could be handled in @register_sharding instead?

Reply:
The main reason I didn't try is because I saw that everything registered via `register_sharding_prop_rule` needs to return an `OutputSharding` ([link](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/sharding_prop.py#L49)), which seems to expect the output of each op to be a tensor or list of tensors ([link](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/op_schema.py#L23)).

We could probably make the logic that invokes the sharding rules more generic so that they can be used to register ops that don't return tensors (cc @wanchaol). Although for metadata ops case, I figured that sharding rules vs. metadata requests were different enough that reasoning about them would be simpler if they were separate (open to either option though)

Reply:
I think the metadata handling is sorta special that we don't need to make them go through sharding propagation, suggested in another comment to make them custom op handlers